qiskit/circuit/commutation_checker.py

```diff
@@ -201,7 +201,7 @@ def _hashable_parameters(params):
     return ("fallback", str(params))
 
 
-_skipped_op_names = {"measure", "reset", "delay"}
+_skipped_op_names = {"measure", "reset", "delay", "initialize", "annotated"}
```

Comment:
This is a comment to get to understand the changes, how does this change relate to the addition of the commutative cancellation pass defaults?

Reply:
There's nothing special about `annotated` in particular, so if these guys are causing problems, it might be an indication that there's larger assumptions about the inputs in the `CommutationChecker` right now that need to be examined.

Reply:
Looking at the failure in the first commit, it's because `Operation` doesn't define a `params` field and that's part of the lookup key for the commutation relationship cache: https://github.com/Qiskit/qiskit/blob/main/qiskit/circuit/commutation_checker.py#L97-L98. We probably should just use a `getattr` there and set an empty default. The larger issue we'll have though is for caching is that the lookup key is based on op name and `annotated` is not unique so we'll be caching bogus entries. So if we want to support annotated operations in the commutation analysis/cancellation we'll either have to change the cache key or skip the caching for `AnnotatedOperations`. I think either way fixing `CommutationChecker`'s support for `AnnotatedOperation` could be a standalone followup PR.

Reply:
I think fixing `CommutationChecker`'s logic so it can run on abstract circuits is probably important to do before we merge this PR, since we'd be moving the pass from its current position of running on physical circuits (where it can typically assume it'll only see `Instruction` subclasses) to one where it'll definitely see abstract `Operation` classes.

That said, I agree those fixes should be in a separate PR.

Reply:
> I agree with these points. Thanks for raising them! :-) Is there a reliable way to check whether an operation is described by a (unitary) matrix (and also to retrieve that matrix), i.e. can we just do hasattr(obj, "to_matrix")? Can retrieving the matrix have a large runtime overhead for an arbitrary Operation? If so we might not be willing to do that for commutation cancellation but maybe the user should decide that.

Yeah we can just do `hasattr(obj, "to_matrix")`, in the general case it should be fast for most gates. This is actually what the `Operator` class does already: https://github.com/Qiskit/qiskit/blob/main/qiskit/quantum_info/operators/operator.py#L105 But there is nothing in practice blocking a particular gate object from implementing an expensive function, but I personally wouldn't worry about that to try and guard against. 

> Should the name of an annotated operation always be "annotated"+base_op.name or should we handle this in the commutation checker? Maybe skipping caching for an abstract operation makes sense...

It would need to be `annotate+base_op.name+modifiers` to be a unique look up key. But I would maybe argue against trying to cache annotate operations, I find it unlikely that in practice we'll have a large number of repeated instances of the same exact annotated operation in a circuit.  I think just doing the matrix multiplication whenever we encounter an annotate op makes the most sense.

Reply:
I hope I addressed these points in PR #11948